build_runner/lib/src/server/hot_reload_client/reloading_manager.dart

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:collection';
+
+import 'package:graphs/graphs.dart' as graphs;
+
+abstract class Module {
+  bool get hasOnDestroy;
+
+  bool get hasOnSelfUpdate;
+
+  bool get hasOnChildUpdate;
+
+  void onDestroy(Map data);
+
+  bool onSelfUpdate(Map data);
+
+  bool onChildUpdate(String childId, Module child, [Map data]);
+}
+
+/// Handles reloading order and hooks invocation
+class ReloadingManager {
+  final Future<Module> Function(String) _reloadModule;
+  final Future<Module> Function(String) _loadModule;
+  final void Function() _reloadPage;
+  final List<String> Function(String moduleId) _moduleParents;
+  final Iterable<String> Function() _allModules;
+
+  final Map<String, int> _moduleOrdering = {};
+  SplayTreeSet<String> _dirty;
+  Completer<void> _running = Completer()..complete();
+
+  int moduleTopologicalCompare(String module1, String module2) {
+    var topological =
+        Comparable.compare(_moduleOrdering[module2], _moduleOrdering[module1]);
+    // If modules are in cycle (same strongly connected component) compare their
+    // string id, to ensure total ordering for SplayTreeSet uniqueness.
+    return topological != 0 ? topological : module1.compareTo(module2);
+  }
+
+  void updateGraph() {
+    var allModules = _allModules();
+    var stronglyConnectedComponents = graphs.stronglyConnectedComponents(
+        allModules, (x) => x, _moduleParents);
+    _moduleOrdering.clear();
+    for (var i = 0; i < stronglyConnectedComponents.length; i++) {
+      for (var module in stronglyConnectedComponents[i]) {
+        _moduleOrdering[module] = i;
+      }
+    }
+  }
+
+  ReloadingManager(this._reloadModule, this._loadModule, this._reloadPage,
+      this._moduleParents, this._allModules) {
+    _dirty = SplayTreeSet(moduleTopologicalCompare);
+  }
+
+  Future<void> run(List<String> modules) async {
+    _dirty.addAll(modules);
+
+    // As function is async, it can potentially be called second time while
+    // first invocation is still running. In this case just mark as dirty and
+    // wait until loop from the first call will do the work
+    if (!_running.isCompleted) return await _running.future;
+    _running = Completer();
+
+    while (_dirty.isNotEmpty) {
+      var moduleId = _dirty.first;
+      _dirty.remove(moduleId);
+
+      var existing = await _loadModule(moduleId);
+      Map data;
+      if (existing.hasOnDestroy) {
+        data = {};
+        existing.onDestroy(data);
+      }
+
+      var newVersion = await _reloadModule(moduleId);
+      bool success;
+      if (newVersion.hasOnSelfUpdate) {
+        success = newVersion.onSelfUpdate(data);
+      }
+      if (success == true) continue;
```

Comment:
How about we change this to return an Enum describing the possible states instead of treating `null` in a special way. In general we try to avoid apis that have a return type but might actually return `null` in Dart (interestingly I don't see a line item about this in effective dart, but its not a common pattern).

Reply:
I thought about it. But where we should declare this enum to use it in modules, and how to access to it from raw js code in bootstap?

Reply:
Since you need access on the JS side I think you would have to make an enum-like object in JS, and then you can access that via package:js.

Essentially, in JS you would have some global object available that looks like this:

```
someGlobalScope.MyEnum = { 'bar': Symbol('bar'), 'baz': Symbol('baz') };
```

And then you can convert that to a Dart class with static getters like this:

```
@JS('someGlobalScope.MyEnum ')
abstract class MyEnum {
  external static int get bar;
  external static int get baz;
}
```

Reply:
Sounds good, but where to place this JS. It will be a part of public api, so I feel it a bad idea to put it inside bootstrap file

Reply:
We should be able to inject it in the same place we inject the other hot reloading code (based on the entrypoint marker?)